polychrom_hoomd/forces.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+from hoomd import md
+
+
+def kg_func(theta, kappa):
+    """Kremer-Grest bending energy penalty function"""
+
+    V = kappa * (1. + np.cos(theta))
+    T = kappa * np.sin(theta)
+
+    return V, T
+    
+    
+def poly_func(r, rmin, rmax, epsilon):
+    """Polychrom soft-core repulsion"""
+
+    term = (r* np.sqrt(6/7)) / rmax
+    
+    V = epsilon * (1 + (term**12) * (term**2 - 1) * (823543./46656.))
+    F = -epsilon * (12.0*r**13/rmax**14 + 84.0*r**11*(0.857142857142857*r**2/rmax**2 - 1)/rmax**12)
+    
+    return V, F
+
+
+def DPD_func(r, rmin, rmax, epsilon):
+    """DPD soft-core repulsion"""
+    
+    V = epsilon * rmax * (1. - r/rmax)**2
+    F = epsilon * (1. - r/rmax)
+    
+    return V, F
+    
+    
+def CSW_func(r, rmin, rmax, epsilon, rcut, n=2500, m=2500):
+    """Continuous square-well potential (see https://doi.org/10.1080/00268976.2018.1481232)"""
+    
+    term = np.exp(-m*(r-rmin)*(r-rcut))
+    
+    V = epsilon/2. * ((rmin/r)**n + (1.-term)/(1.+term) - 1)
+    F = epsilon * (n/2.*(rmin/r)**(n+1) - m*(2*r-rcut-rmin)*term/(1.+term)**2)
+    
+    return V, F
+
+
+def set_excluded_volume(nlist, mode_integ,
+                        seed=0,
+                        epsilon_exc=10,
+                        mode_exc="dpd",
+                        **kwargs):
+    """Set (soft) excluded-volume repulsion based on choice of thermostat"""
+
+    if mode_integ == "langevin":
+        poly_exc = md.pair.table(width=1000, nlist=nlist, name="exc")
+        poly_exc.pair_coeff.set(['A', 'B'], ['A', 'B'],
+                                 func=DPD_func if mode_exc == "dpd" else poly_func,
+                                 rmin=0., rmax=1., coeff=dict(epsilon=epsilon_exc))
+                                 
+    elif mode_integ == "dpd":
+        poly_exc = md.pair.dpd(nlist=nlist, kT=1.0, r_cut=1.0, seed=seed)
+        poly_exc.pair_coeff.set(['A', 'B'], ['A', 'B'],
+                                A=epsilon_exc, gamma=1)
+                                
+    return poly_exc
+
+
+def set_PSW_attraction(nlist,
+                       epsilon_att=0.20, r_att=1.5, pad=1e-2,
+                       **kwargs):
```

Comment:
I wonder if there is an intuitive way to expose the groups over which the attactions are applied... e.g. as an interaction matrix ?